content/posts/numpy/numpy-rng/index.md

```diff
@@ -0,0 +1,136 @@
+
+From NumPy 1.17, it is now very easy to instantiate independent RNGs. Depending on the RNG that you use, different strategies are easily available as documented in the [Parallel random number generation section](https://numpy.org/doc/stable/reference/random/parallel.html) of the NumPy documentation. One of the strategies is to use `SeedSequence` which is an algorithm that makes sure that a not so good user-provided seed results in a good initial state for the RNG. Additionally, it ensures that two close seeds will result in two very different initial states for the RNG that are, with very high probability, independent of each other. You can refer to the documentation of [SeedSequence Spawning](https://numpy.org/doc/stable/reference/random/parallel.html#seedsequence-spawning) for an example on how to generate independent RNGs from a user-provided seed. I here show how you can do this from an already existing RNG and apply it to the [joblib example](https://joblib.readthedocs.io/en/latest/auto_examples/parallel_random_state.html#fixing-the-random-state-to-obtain-deterministic-results) mentioned above.
+
+
+```python
+import numpy as np
+from joblib import Parallel, delayed
+
+
+def stochastic_function(seed, high=10):
+    rng = np.random.default_rng(seed)
+    return rng.integers(high, size=5)
+
+
+seed = 319929794527176038403653493598663843656
+# create the RNG that you want to pass around
+rng = np.random.default_rng(seed)
+# get the SeedSequence of the passed RNG.
+# WARNING: note that this is using a private attribute of
+# the RNG and the API might change in the future.
+ss = rng.bit_generator._seed_seq
+# create 5 initial independent states
+child_states = ss.spawn(5)
+
+# use 2 processes to run the stochastic_function 5 times with joblib
+random_vector = Parallel(n_jobs=2)(
+    delayed(stochastic_function)(random_state) for random_state in child_states
+)
+print(random_vector)
+
+# rerun to check that we obtain the same outputs
+random_vector = Parallel(n_jobs=2)(
+    delayed(stochastic_function)(random_state) for random_state in child_states
+)
+print(random_vector)
+```
+
+By using a fixed seed you always get the same results and by using `SeedSequence.spawn` you have an independent RNG for each of the iterations. Note that I used the convenient `default_rng` function in `stochastic_function`. You can also see that the `SeedSequence` of the existing RNG is a private attribute. Accessing the `SeedSequence` might become easier in future versions of NumPy (more information [here](https://github.com/numpy/numpy/issues/15322#issuecomment-626400433)).
+
```

Comment:
The end feels a bit abrupt; it may be helpful to have a short conclusion here. It could simply say that the new random API provides safer ways to generate near-independent random streams of numbers, and that the resources below are available for further reading.